typescript/18-wordy/wordy.ts

```diff
@@ -0,0 +1,146 @@
+type ValidToken =
+  | { type: 'OPERAND'; value: number }
+  | { type: 'OPERATOR'; value: string };
+
+type Token =
+  | ValidToken
+  | { type: 'UNKNOWN'; value: string };
+
+class Node {
+  operator: string;
+  leftOperand: Node | number;
+  rightOperand: Node | number;
+
+  constructor(operator: string, leftOperand: Node | number, rightOperand: Node | number) {
+    this.operator = operator;
+    this.leftOperand = leftOperand;
+    this.rightOperand = rightOperand;
+  }
+}
+
+function preValidateTokens(tokens: Token[]): ValidToken[] {
+  tokens.forEach((val, i) => {
+    if (val.type === 'UNKNOWN') throw new Error('Unknown operation')
+    if (i > 0 && (val.type === tokens[i - 1].type || tokens.length < 3)) throw new Error('Syntax error')
+  })
+
+  return tokens as ValidToken[];
+}
+
+function buildAST(tokens: ValidToken[], priority: Record<string, number>): Node | number {
+  const trees: (Node | number)[] = [];
+  const output: (ValidToken | number)[] = [];
+  const operators: ValidToken[] = [];
+
+  for (let i = 0; i < tokens.length; i++) {
+    const token: Token = tokens[i];
+
+    if (token.type === 'OPERAND') {
+      output.push(token.value);
```

Comment:
If you imagine this loop as an isolated function, it's signature would essentially be:
```
input:
  { type: 'OPERAND'; value: number } | { type: 'OPERATOR'; value: string }
output:
  number | { type: 'OPERATOR'; value: string }
```
What this loop does is reordering of valid tokens. Extracting operand `value` doesn't really do anything useful here, just adds complexity and confusion to the type system, since instead of checking `type == 'OPERAND'` you now have to check `typeof item === 'number'` further in the code.
Fix the types to benefit you.